src/nspawn/nspawn.c

```diff
@@ -2218,6 +2285,29 @@ static int setup_machine_id(const char *directory) {
         return 0;
 }
 
+static int recursive_chown(const char *directory, uid_t shift, uid_t range) {
+        int r;
+
+        assert(directory);
+
+        if (arg_userns_mode == USER_NAMESPACE_NO || !arg_userns_chown)
+                return 0;
+
+        r = path_patch_uid(directory, arg_uid_shift, arg_uid_range);
+        if (r == -EOPNOTSUPP)
+                return log_error_errno(r, "Automatic UID/GID adjusting is only supported for UID/GID ranges starting at multiples of 2^16 with a range of 2^16.");
+        if (r == -EBADE)
+                return log_error_errno(r, "Upper 16 bits of root directory UID and GID do not match.");
+        if (r < 0)
+                return log_error_errno(r, "Failed to adjust UID/GID shift of OS tree: %m");
```

Comment:
I also don't really understand the strace output. It gives consistently the same output though.

rkt pre-mounts `/sys` because it pre-mounts `/sys/fs/cgroup` and we cannot pre-mount the latter if the former is not also pre-mounted the former. We pre-mount the cgroups with write access to some knob files to allow systemd in the container to enforce the limits on some apps of the container:
```
[Service]
MemoryLimit=...
CPUQuota=...
```

(otherwise, `systemd-nspawn` does not give write access to anything in the memory/cpu cgroup subsystem)